### PR TITLE
Fix #5005: Reorganize CODEOWNERS for clearer segmentation & to use GitHub groups

### LINKED ISSUE #5005: Identifying the groups of codeowners to organize (the current assumption is that this will roughly match the existing Android teams)

### PR DESCRIPTION
**Pull request by BenHenning**
## Explanation
Fixes #5005.
Fixes part of #5004.

This PR completed reorganizes codeowners in two ways:
1. It tries to more aggressively segment directories based on common themes. Longer term (per #5004) the goal is to eventually organize the codebase itself based on features and infrastructural components, but that requires much more that needs to be done yet.
2. It moves all codeownership over to four teams: @oppia/owners, @oppia/android-app-infrastructure-reviewers, @oppia/android-dev-workflow-reviewers, and @oppia/android-frontend-reviewers

This PR does attempt to keep the divisions more logical than necessarily even. At a high-level:
- oppia/owners own anything that's especially sensitive and requires additional care or should never really be changed.
- oppia/android-frontend-reviewers own everything tied to code that directly affects the end-user experience.
- oppia/android-app-infrastructure-reviewers own everything that both supports and defines domain logic.
- oppia/android-dev-workflow-reviewers own everything that supports making development easier, or is tied to central developer workflows (such as the build infrastructure).

The above don't exactly match to existing teams. However, it's generally assumed that:
- The team TL will generally review anything that goes to oppia/owners.
- The CLaM team is ultimately responsible for everything that falls under oppia/android-frontend-reviewers and oppia/android-app-infrastructure-reviewers. The distinction between these two types of parts of the codebase allow for more granular specialization as the knowledge required in each of these coding domains tends to be different.
- The developer workflow team is ultimately responsible for everything that falls under oppia/android-dev-workflow-reviewers-reviewers.

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
N/A -- this PR is only changing codeowners (which doesn't affect the actual app or any builds/configurations).

### CODE REVIEW

**Comment by BenHenning** [comment]
@seanlip PTAL for approval.

@adhiamboperes PTAL for thoughts on the organization & distribution between the different reviewer groups. Please note the part of the PR description that mentions which areas are expected to fall under the CLaM team.

File: .github/CODEOWNERS
Comment (seanlip) [inline .github/CODEOWNERS]: Can we use codeowner teams instead for this first paragraph? I don't think we should be doing issue filing etc. here, or editing the codeowner file -- that is much too clunky IMO. The procedure should be something like this instead:

(a) make sure there are other folks on your codeowner team
(b) send a note to them and make sure they can cover, and tell them when you'll be back; include Ben/Sean/Adhiambo on that note
(c) update your status to "busy" to make sure you don't get reviews
(d) when you're back, change your status back so that you get new reviews

File: .github/CODEOWNERS
Comment (seanlip) [inline .github/CODEOWNERS]: Can we call this android-dev-workflow-reviewers instead, to (a) increase distinguishability from the "app infrastructure" team, and (b) mimic Web's "dev workflow" team naming? (It's also shorter, which would make @-mentioning the team a little more readable, but that's a minor benefit.)

File: .github/CODEOWNERS
Comment (seanlip) [inline .github/CODEOWNERS:184]: Wouldn't these all be dev-workflow, like the stuff under "sensitive files" below? I would reserve `owners` for things that only the repo admins should modify, and I think that the dev workflow team should totally have discretion to change stuff in this section based on developers' needs.

It might be that "owners" is a temporary substitute for @seanlip, since I can review dev workflow stuff but not dev tooling. In that case, I think the overall category should still be `android-dev-workflow-reviewers`, but for now you might want to temporarily have an `android-devtools-reviewers` team for the bazel/kotlin stuff. Later, when things are stable enough for the dev workflow team to include those in their remit (since they should be ensuring the build,/onboarding/etc. processes are streamlined for contributors), we can combine those into a single `android-dev-workflow-reviewers` team.

**Review by seanlip — CHANGES_REQUESTED** [review]
Thanks @BenHenning -- had a few thoughts. Please see comments.

File: .github/CODEOWNERS
Comment (BenHenning) [reply .github/CODEOWNERS]: Ah, that's what the new comment was meant to suggest (since I removed the previous instructions that explained adding a TODO). These steps are much more actionable though, so I'll switch it over to these--thanks!

Comment (seanlip) [reply .github/CODEOWNERS]: Sg, though just to confirm, the updated comment in the PR still had "please replace your ownership with a delegate, file an issue" in it. There's no need to file an issue, right?

Comment (BenHenning) [reply .github/CODEOWNERS]: Ah, yes that's correct. An issue isn't needed.

File: .github/CODEOWNERS
Comment (BenHenning) [reply .github/CODEOWNERS]: Seems reasonable, done!

File: .github/CODEOWNERS
Comment (BenHenning) [reply .github/CODEOWNERS:184]: This wasn't meant to be a substitute for you, actually. The thinking here is that these files should rarely change and, when they do, there should be a good reason that the team's technical leadership approve. We're mainly talking about the README, PR/issue templates, and .gitignore. All of these files need to be carefully reviewed and often require context that exceeds the purview of any of the individual reviewer teams.

That being said, I'm not opposed with making the README & gitignore special one-offs for owners reviews, and making everything else android-dev-workflow-reviewers since I think it's not unreasonable for them to own the issue & discussion templates.

It also occurred to me that .github/ was not meant to be as much of a blanket pattern as it is here, so that's been fixed as well. Hopefully this actually better helps clarify the distinction between these two types of files.

**Comment by BenHenning** [comment]
Thanks @seanlip! I attempted to address all of your comments--PTAL.

File: .github/CODEOWNERS
Comment (seanlip) [inline .github/CODEOWNERS]: "the all" --> "all"

File: .github/CODEOWNERS
Comment (seanlip) [inline .github/CODEOWNERS]: Suggest: ...that should only be changed after careful consideration.

(I think of "carefully" as relating to the act of making the changes, rather than the content of the changes themselves.)

File: .github/CODEOWNERS
Comment (seanlip) [inline .github/CODEOWNERS]: Ditto.

**Review by seanlip — APPROVED** [review]
Thanks, the new distinction for the /.github files seems fine. Approving to unblock, but PTAL at the new comments before merging. Thanks!

File: .github/CODEOWNERS
Comment (BenHenning) [reply .github/CODEOWNERS:184]: No follow-up comment and you mentioned LGTM @seanlip, so resolving this thread.

File: .github/CODEOWNERS
Comment (BenHenning) [reply .github/CODEOWNERS]: No follow-up comment and you mentioned LGTM @seanlip, so resolving this thread.

File: .github/CODEOWNERS
Comment (BenHenning) [reply .github/CODEOWNERS]: No follow-up comment and you mentioned LGTM @seanlip, so resolving this thread.

File: .github/CODEOWNERS
Comment (BenHenning) [reply .github/CODEOWNERS]: Done.

File: .github/CODEOWNERS
Comment (BenHenning) [reply .github/CODEOWNERS]: Good call--done.

File: .github/CODEOWNERS
Comment (BenHenning) [reply .github/CODEOWNERS]: Done.

Comment (BenHenning) [reply .github/CODEOWNERS]: I've verified this change has landed in the PR, so marking this as resolved since you've approved the PR @seanlip.

File: .github/CODEOWNERS
Comment (BenHenning) [reply .github/CODEOWNERS]: I've verified this change has landed in the PR, so marking this as resolved since you've approved the PR @seanlip.

File: .github/CODEOWNERS
Comment (BenHenning) [reply .github/CODEOWNERS]: I've verified this change has landed in the PR, so marking this as resolved since you've approved the PR @seanlip.

**Comment by BenHenning** [comment]
Enabling auto-merge.

@adhiamboperes please let Sean or I know if you have any follow-up concerns or suggestions for the codeowners changes.

**Comment by seanlip** [comment]
Confirming that the new changes LGTM. Thanks!

**Pull request merged**
